README.md
# xr_robot

小R机器人小车APP应用示例。包括基础组件components和Demo应用两部分。

## 基础组件
基础组件components中包含机器人运动控制中间件xrrobot package和运动控制消息xrrobot_msgs package。

## Demo应用
Demo应用包括
\- gesture_control：通过手势控制机器人小车运动，包括左右旋转和前后平移运动。
\- body_tracking：机器人小车跟随人体移动。


body_tracking/CHANGELOG.md
# Changelog for package hobot_app_xrrobot_body_tracking

tros_2.0.0 (2023-05-11)
------------------
1. 更新package.xml，支持应用独立打包


components/xrrobot/src/XRMiddleware.cpp
// Copyright (c) 2022，Horizon Robotics.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <termios.h>
#include <unistd.h>

#include <thread>

#include "xr_middleware.c"

class XRMiddleware {
 public:
  XRMiddleware(const char *p, int s) {
    port_ = p;
    speed_ = s;
  }

  int Init() {
    int ret = -1;
    fd_ = open(port_, O_RDWR | O_NOCTTY | O_NDELAY);  //打开串口设备
    if (fd_ < 0) {
      printf("open xrbase failure:%s\r\n", strerror(errno));
      return -1;
    }

    set_serial(fd_, speed_, 8, 'N', 1);
    printf("open port successfully: %s, fd: %d\n", port_, fd_);
    tcflush(fd_, TCIOFLUSH);  //  清空串口缓存
    //  usleep(20000);
    //  ret = _XRMiddleWare_deinit(fd_);
    //  usleep(20000);
    //  ret = _XRMiddleWare_clrOdom(fd_);
    //  usleep(20000);
    ret = _XRMiddleWare_init(fd_);
    usleep(20000);
    ret = _XRMiddleWare_init(fd_);
    usleep(20000);
    ansythread = std::make_shared<std::thread>([=]() { _XR_readBytes(fd_); });
    return ret;
  }

  int Deinit() {
    int ret = -1;
    ret = _XRMiddleWare_deinit(fd_);
    usleep(20000);
    ret = _XRMiddleWare_clrOdom(fd_);
    usleep(20000);
    ret = _XRMiddleWare_LowPower(fd_);
    usleep(3 * 1000 * 1000);
    return ret;
  }

  int SetPID(u_int16_t P, u_int16_t I, u_int16_t D) {
    return _XRMiddleWare_SetPID(fd_, P, I, D);
  }

  int SetParams(u_int16_t encoder_resolution,
                float wheel_radius,
                float wheel_track_div2,
                float half_long) {
    auto ret = _XRMiddleWare_set_encoder_resolution(fd_, encoder_resolution);
    usleep(20000);
    ret = _XRMiddleWare_set_wheel_diameter(fd_, wheel_radius);
    usleep(20000);
    ret = _XRMiddleWare_set_wheel_track(fd_, wheel_track_div2);
    usleep(20000);
    ret = _XRMiddleWare_set_half_long(fd_, half_long);
    return ret;
  }

  int SetVelocity(float x, float y, float yaw) {
    return _XRMiddleWare_SetVelocity(fd_, x, y, yaw);
  }

  int SetWheelspeed_(u_int16_t w1, u_int16_t w2, u_int16_t w3, u_int16_t w4) {
    return _XRMiddleWare_SetWheelSpeed(fd_, w1, w2, w3, w4);
  }

  int GetOdom(float *odom_buf_6) { return _XRMiddleWare_GetOdom(odom_buf_6); }

  int ClrOdom() { return _XRMiddleWare_clrOdom(fd_); }

  int GetIMU(float *imu_buf_9) { return _XRMiddleWare_GetIMU(imu_buf_9); }

  int GetWheelspeed_(u_int16_t *wheelspeed__buf_8) {
    return _XRMiddleWare_GetWheelSpeed(wheelspeed__buf_8);
  }

  float GetBattery() { return _XRMiddleWare_GetBattery(); }

  int LowPwr() { return _XRMiddleWare_LowPower(fd_); }

 private:
  int fd_ = -1;
  const char *port_ = nullptr;
  int speed_ = 115200;
  std::shared_ptr<std::thread> ansythread = nullptr;
};


components/xrrobot/src/xrrobot_base.cpp
// Copyright (c) 2022，Horizon Robotics.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Transform.h>
#include <tf2_ros/transform_broadcaster.h>

#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "XRMiddleware.cpp"
#include "nav_msgs/msg/odometry.hpp"
#include "rclcpp/rclcpp.hpp"
#include "xrrobot_msgs/msg/battery.hpp"
#include "xrrobot_msgs/msg/imu.hpp"

using namespace std::chrono_literals;

class XRMiddleWare : public rclcpp::Node {
 public:
  XRMiddleWare() : Node("xr_middle_ware"), count_(0) {
    auto qos = rclcpp::QoS(rclcpp::KeepLast(10));

    battery_pub_ =
        this->create_publisher<xrrobot_msgs::msg::Battery>("voltage", qos);
    imu_pub_ = this->create_publisher<xrrobot_msgs::msg::Imu>("raw_imu", qos);
    odom_pub_ =
        this->create_publisher<nav_msgs::msg::Odometry>("raw_odom", qos);

    update_timer_ = this->create_wall_timer(
        25ms, std::bind(&XRMiddleWare::timer_callback, this));

    cmd_vel_sub_ = this->create_subscription<geometry_msgs::msg::Twist>(
        "cmd_vel",
        qos,
        std::bind(&XRMiddleWare::command_velocity_callback,
                  this,
                  std::placeholders::_1));

    xr = std::make_shared<XRMiddleware>("/dev/xrbase", 115200);
    // xr->Deinit();
    // usleep(20000);
    // xr->SetPID(20, 0, 0);
    // usleep(20000);
    // xr->SetParams(3150, 0.022, 0.0545, 0.051);
    usleep(20000);
    xr->Init();
    usleep(20000);
  }
  void InitTf() {
    tf_broadcaster_ =
        std::make_shared<tf2_ros::TransformBroadcaster>(shared_from_this());
  }

 private:
  void handle_imu() {
    float imu_list[9] = {0};
    xr->GetIMU(imu_list);

    xrrobot_msgs::msg::Imu imu_data;
    imu_data.angular_velocity.x = imu_list[0];
    imu_data.angular_velocity.y = imu_list[1];
    imu_data.angular_velocity.z = imu_list[2];
    //  printf("imu z: %f\n", imu_list[2]);
    imu_data.linear_acceleration.x = imu_list[3];
    imu_data.linear_acceleration.y = imu_list[4];
    imu_data.linear_acceleration.z = imu_list[5];
    imu_data.magnetic_field.x = imu_list[6];
    imu_data.magnetic_field.y = imu_list[7];
    imu_data.magnetic_field.z = imu_list[8];
    imu_pub_->publish(imu_data);
  }

  void handle_odom() {
    float factor = 1.0f;
    float odom_list[6] = {0};
    xr->GetOdom(odom_list);

    geometry_msgs::msg::TransformStamped tf_transfer;
    nav_msgs::msg::Odometry odom;
    odom.header.stamp = this->now();
    odom.header.frame_id = "odom";
    odom.child_frame_id = "base_footprint";
    odom.pose.pose.position.x = odom_list[0] / factor;
    odom.pose.pose.position.y = odom_list[1] / factor;
    //  printf("odom p: [x, y]: [%f, %f]\n", odom_list[0], odom_list[1]);
    odom.pose.pose.position.z = 0;

    tf2::Quaternion q;
    q.setRPY(0, 0, odom_list[2] / factor);
    //  printf("odom yaw: %f\n", odom_list[2]);
    odom.pose.pose.orientation.x = q.x();
    odom.pose.pose.orientation.y = q.y();
    odom.pose.pose.orientation.z = q.z();
    odom.pose.pose.orientation.w = q.w();

    // printf("odom orientation: [x, y, z, w]: [%f, %f, %f, %f]\n",
    // 		    q.x(), q.y(), q.z(), q.w());

    odom.twist.twist.linear.x = odom_list[3] / factor;
    odom.twist.twist.linear.y = odom_list[4] / factor;
    odom.twist.twist.angular.z = odom_list[5] / factor;
    // printf("odom linear: [x, y, angularz]: [%f, %f, %f]\n",
    // 		    odom_list[3], odom_list[4], odom_list[5]);
    odom.pose.covariance[0] = 0.001;
    odom.pose.covariance[7] = 0.001;
    odom.pose.covariance[35] = 0.001;
    odom.twist.covariance[0] = 0.0001;
    odom.twist.covariance[7] = 0.0001;
    odom.twist.covariance[35] = 0.0001;

    odom_pub_->publish(odom);

    tf_transfer.header.stamp = odom.header.stamp;
    tf_transfer.header.frame_id = odom.header.frame_id;
    tf_transfer.child_frame_id = odom.child_frame_id;
    tf_transfer.transform.translation.x = odom.pose.pose.position.x;
    tf_transfer.transform.translation.y = odom.pose.pose.position.y;
    tf_transfer.transform.translation.z = odom.pose.pose.position.z;
    tf_transfer.transform.rotation = odom.pose.pose.orientation;

    //  tf_broadcaster_->sendTransform(tf_transfer);
  }

  void handle_bat() {
    auto battery_data = xr->GetBattery();
    xrrobot_msgs::msg::Battery bat;
    bat.battery = battery_data;
    battery_pub_->publish(bat);
  }

  void timer_callback() {
    count_++;
    handle_imu();
    handle_odom();
    if (count_ > 60) {
      count_ = 0;
      handle_bat();
    }
  }

  void command_velocity_callback(
      const geometry_msgs::msg::Twist::SharedPtr cmd_vel_msg) {
    //	  printf("set z: %f\n", cmd_vel_msg->angular.z);
    xr->SetVelocity(cmd_vel_msg->linear.x, 0, cmd_vel_msg->angular.z);
  }

  rclcpp::TimerBase::SharedPtr update_timer_;

  rclcpp::Publisher<xrrobot_msgs::msg::Battery>::SharedPtr battery_pub_;
  rclcpp::Publisher<xrrobot_msgs::msg::Imu>::SharedPtr imu_pub_;
  rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;

  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_sub_;

  size_t count_;

  std::shared_ptr<XRMiddleware> xr;

  std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
};

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<XRMiddleWare>();
  node->InitTf();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}


gesture_control/launch/hobot_app_xrrobot_gesture_control.launch.py
# Copyright (c) 2022，Horizon Robotics.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from launch import LaunchDescription
from launch_ros.actions import Node

def generate_launch_description():
    return LaunchDescription([
        # 启动图片发布pkg
        # F37 mipi cam
        # Node(
        #     package='mipi_cam',
        #     executable='mipi_cam',
        #     output='screen',
        #     parameters=[
        #         {"out_format": "nv12"},
        #         {"image_width": 960},
        #         {"image_height": 544},
        #         {"io_method": "shared_mem"},
        #         {"video_device": "F37"}
        #     ],
        #     arguments=['--ros-args', '--log-level', 'error']
        # ),
        # usb cam
        Node(
            package='hobot_usb_cam',
            executable='hobot_usb_cam',
            output='screen',
            parameters=[
                {"image_width": 640},
                {"image_height": 480},
                {"video_device": "/dev/video8"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),
        # 启动jpeg图片编码&发布pkg
        Node(
            package='hobot_codec',
            executable='hobot_codec_republish',
            output='screen',
            parameters=[
                {"channel": 1},
                {"in_mode": "ros"},
                {"in_format": "jpeg"},
                {"out_mode": "shared_mem"},
                {"out_format": "nv12"},
                {"sub_topic": "/image"},
                {"pub_topic": "/hbmem_img"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),
        # 启动单目rgb人体、人头、人脸、人手框和人体关键点检测pkg
        Node(
            package='mono2d_body_detection',
            executable='mono2d_body_detection',
            output='screen',
            parameters=[
                {"ai_msg_pub_topic_name": "/hobot_mono2d_body_detection"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),
        # 启动人手关键点检测pkg
        Node(
            package='hand_lmk_detection',
            executable='hand_lmk_detection',
            output='screen',
            parameters=[
                {"ai_msg_sub_topic_name": "/hobot_mono2d_body_detection"},
                {"ai_msg_pub_topic_name": "/hobot_hand_lmk_detection"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),
        # 启动web展示pkg
        Node(
            package='websocket',
            executable='websocket',
            output='screen',
            parameters=[
                {"image_topic": "/image"},
                {"image_type": "mjpeg"},
                {"smart_topic": "/hobot_hand_gesture_detection"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),
        # 启动手势识别pkg
        Node(
            package='hand_gesture_detection',
            executable='hand_gesture_detection',
            output='screen',
            parameters=[
                {"ai_msg_sub_topic_name": "/hobot_hand_lmk_detection"},
                {"ai_msg_pub_topic_name": "/hobot_hand_gesture_detection"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),
        # 启动手势交互pkg
        Node(
            package='gesture_control',
            executable='gesture_control',
            output='screen',
            parameters=[
                {"ai_msg_sub_topic_name": "/hobot_hand_gesture_detection"},
                {"twist_pub_topic_name": "/cmd_vel"},
                {"activate_wakeup_gesture": 1},
                {"track_serial_lost_num_thr": 100},
                {"move_step": 0.5},
                {"rotate_step": 0.5}
            ],
            arguments=['--ros-args', '--log-level', 'warn']
        ),
        
        # 启动小R机器⼈小⻋pkg
        Node(
            package='xrrobot',
            executable='xrrobot',
            output='screen',
            arguments=['--ros-args', '--log-level', 'info']
        )
    ])


body_tracking/launch/hobot_app_xrrobot_body_tracking.launch.py
# Copyright (c) 2022，Horizon Robotics.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from launch import LaunchDescription
from launch_ros.actions import Node

def generate_launch_description():
    return LaunchDescription([
        # 启动图片发布pkg
        # F37 mipi cam
        # Node(
        #     package='mipi_cam',
        #     executable='mipi_cam',
        #     output='screen',
        #     parameters=[
        #         {"out_format": "nv12"},
        #         {"image_width": 960},
        #         {"image_height": 544},
        #         {"io_method": "shared_mem"},
        #         {"video_device": "F37"}
        #     ],
        #     arguments=['--ros-args', '--log-level', 'error']
        # ),
        # usb cam
        Node(
            package='hobot_usb_cam',
            executable='hobot_usb_cam',
            output='screen',
            parameters=[
                {"image_width": 640},
                {"image_height": 480},
                {"video_device": "/dev/video8"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),

        # 启动jpeg图片编码&发布pkg
        Node(
            package='hobot_codec',
            executable='hobot_codec_republish',
            output='screen',
            parameters=[
                {"channel": 1},
                {"in_mode": "ros"},
                {"in_format": "jpeg"},
                {"out_mode": "shared_mem"},
                {"out_format": "nv12"},
                {"sub_topic": "/image"},
                {"pub_topic": "/hbmem_img"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),
        
        # 启动单目rgb人体、人头、人脸、人手框和人体关键点检测pkg
        Node(
            package='mono2d_body_detection',
            executable='mono2d_body_detection',
            output='screen',
            parameters=[
                {"ai_msg_pub_topic_name": "/hobot_mono2d_body_detection"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),
        # 启动人手关键点检测pkg
        Node(
            package='hand_lmk_detection',
            executable='hand_lmk_detection',
            output='screen',
            parameters=[
                {"ai_msg_sub_topic_name": "/hobot_mono2d_body_detection"},
                {"ai_msg_pub_topic_name": "/hobot_hand_lmk_detection"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),
        # 启动web展示pkg
        Node(
            package='websocket',
            executable='websocket',
            output='screen',
            parameters=[
                {"image_topic": "/image"},
                {"image_type": "mjpeg"},
                {"smart_topic": "/hobot_hand_gesture_detection"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),
        # 启动手势识别pkg
        Node(
            package='hand_gesture_detection',
            executable='hand_gesture_detection',
            output='screen',
            parameters=[
                {"ai_msg_sub_topic_name": "/hobot_hand_lmk_detection"},
                {"ai_msg_pub_topic_name": "/hobot_hand_gesture_detection"}
            ],
            arguments=['--ros-args', '--log-level', 'error']
        ),
        # 启动手势交互pkg
        Node(
            package='body_tracking',
            executable='body_tracking',
            output='screen',
            parameters=[
                {"ai_msg_sub_topic_name": "/hobot_hand_gesture_detection"},
                {"twist_pub_topic_name": "/cmd_vel"},
                {"activate_wakeup_gesture": 0},
                {"img_width": 960},
                {"img_height": 544},
                {"track_serial_lost_num_thr": 100},
                {"move_step": 0.3},
                {"rotate_step": 0.5},
                {"activate_robot_move_thr": 5},
                {"activate_robot_rotate_thr": 45}
            ],
            arguments=['--ros-args', '--log-level', 'warn']
        ),

        # 启动小R机器⼈小⻋
        Node(
            package='xrrobot',
            executable='xrrobot',
            output='screen',
            arguments=['--ros-args', '--log-level', 'warn']
        )
    ])
